### DISCUSSION TITLE
<script> tag not working on page change, only on initial load and reload

### DISCUSSION THREAD

**Discussion by junaid33** [vercel/next.js#17919] [Help] [answered]
Hello,

I have 2 3rd party script tags in _document.js. The script interacts with divs on the page based on classNames. The script works perfectly on the initial load but stops working if I navigate to other pages. If I refresh the page, the script starts working again. 

This is an ongoing issue: https://github.com/vercel/next.js/issues/4477

Anyway to get the script tag to work on page change?

**Comment by jamesmosier** [comment, marked as answer]
Hi @junaid33. Do you happen to have an example of what you are seeing? 

This sounds similar to another issue where `dangerouslySetInnerHTML ` was not working properly when doing client side navigation due to the manner in which `dangerouslySetInnerHTML ` works: https://github.com/vercel/next.js/discussions/17346#discussioncomment-84584

**Reply by junaid33** [reply]
Yes here a simple reproduction: https://github.com/junaid33/nextjs-script-error

I am not using `dangerouslySetInnerHTML` in any of the code.

**Reply by jamesmosier** [reply]
The issue is that the script is already initialized on the first page load and attached to any DOM elements in order to interact with them. When you navigate to a new page, the script would need to be reinitialized either internally (in the 3rd party script) or via your own code.

One way around this in your own code would be to use `next/head` to import it on each page:

```js
import Head from 'next/head'

<Head>
  <script src="your-3rd-party-script.js"></script>
</Head>
```

**Reply by junaid33** [reply]
I tried that also but it doesn't work. I pushed it to repo if you'd like to see.

**Reply by clementkelly** [reply]
Hey guys, I've got the same issue on my side. I've got a few different Script Tags loaded into the Head (Next/Head) of my various pages but they won't trigger when changing pages using Next/Link. Did you guys have an update/solution on this issue?

**Reply by mdinos** [reply]
Not sure why this has been marked as answered. I'm having the same issue, I had the solution that @jamesmosier suggested already implemented in my code already.

**Reply by pdebruic** [reply]
Yea no kidding.  This isn't answered.

**Reply by surjithctly** [reply]
Same issue :(

Did you guys manage to solve it? If yees, how? Any workaround?

**Reply by junaid33** [reply]
Honestly, I could not find a fix so just reverse engineered the script. Normally it makes a fetch call and manipulates the data so replicating that should not be too hard. In my case, the script tag was sending actual HTML. I used Cheerio in the getStaticProps function to parse the data from the HTML and pass it to the page. It might be too much depending on your script tag, but script tags do not persist using shallow linking (Link component from Next). Another option is to use actual <a /> tags and then the script tag is run. Again this might work, just depends on what the script tag code is doing. Downside to the former option is that if the script tag API or HTML changes, you will have to update as well.

**Reply by japhetkuntu** [reply]
yeah, I found a solution [here](https://stackoverflow.com/questions/34424845/adding-script-tag-to-react-jsx)

**Reply by ValentinDuboscq** [reply]
Hi guys, the only workaround I found was to use `<a>` tag instead of `<Link>` but I don't want to do that.

Any answer ?

**Reply by BluetyIt** [reply]
Same problems, the script doesn't run after page navigation.

**Reply by Oluwaseun4244** [reply]
> Hi guys, the only workaround I found was to use `<a>` tag instead of `<Link>` but I don't want to do that.
> 
> Any answer ?

exactly, that is the only workaround, it kind of negate the one page feature of react, because "a" tag takes href. hence, the page refreshes upon navigation. let me know if you later found the solution.

**Reply by narges22** [reply]
I found a solution by adding an id to the script tag, after client navigation we can check the existence of the element.
here is the example code:
` <Scripts id="externalJs" src="https://js.tuna.uy/tuna.js" onLoad={loadScriptHandler} />`
```
useEffect(() => {
        const existingScript = document.getElementById('externalJs');
        if (existingScript) {
            loadScriptHandler();
        }
    }, [router.asPath]);
```

**Reply by clearly-outsane** [reply]
What does loadScriptHandler do here?

**Reply by jinnn0** [reply]
@bobywoodwarrior same for me. But I actually had to use `<a>` instead of `'next/link' `as a temp solution. Is this thread resolved at this point at all ?

**Reply by jinnn0** [reply]
@narges22 could you please add the code for `loadScriptHandler` as well

**Reply by talhaa99** [reply]
> The issue is that the script is already initialized on the first page load and attached to any DOM elements in order to interact with them. When you navigate to a new page, the script would need to be reinitialized either internally (in the 3rd party script) or via your own code.
> 
> One way around this in your own code would be to use `next/head` to import it on each page:
> 
> ```js
> import Head from 'next/head'
> 
> <Head>
>   <script src="your-3rd-party-script.js"></script>
> </Head>
> ```

Hello! I got into a similar problem and this solution works. Thanks, buddy @jamesmosier

**Reply by dmoKruso** [reply]
@narges22 Interested on checking your `loadScriptHandler` as well. Because everything I tried, it doesn't work.

**Reply by creidinger** [reply]
I rant into this same problem. @talhaa99 your fix worked for me. 

Wish there was a better way to do this.

**Reply by krystiangryczon** [reply]
I had the same problem, but not necessarily when navigating per se, was just trynna use the anchor to scroll down to the section and switching `NextLink` to regular `a` tag worked! And somehow it made the page not refresh, while `NextLink` seemed to refresh the page

**Reply by valourus** [reply]
@krystiangryczon Using a normal `<a>` will do a page refresh, there by losing any state in your current react application. This also reload all .js / .css file and making page switches significantly slower. Using the next/link makes sure the browser doesn't reload and making the React app feel much smoother (and quicker).

**Reply by fvaldez421** [reply]
Hey folks, I was able to find a solution for my use case:
I had to dynamically embed an iframe in my application. It would mount correctly on first load or refresh, but would fail when using next links, and when navigating using browser history forward or backward.

The solution recommended above (using an explicit `<a>` tag) wouldn't work here because the script would fail to reload (and mount) with native browser history fwd/bwd behavior too.

After doing some searching I stumbled upon this [Stackoverflow thread](https://stackoverflow.com/questions/49761241/executing-a-script-tag-on-append-after-the-dom-has-loaded) that recommended the use of `createRange` and `createContextualFragment`. I was able to _React-ify_ it and worked for me, hope it works for you too!

```ts
  const scriptRoot = useRef() // gets assigned to a root node
  const script = `<script>...</script>
  
  useEffect(() => {
    // creates a document range (grouping of nodes in the document is my understanding)
    // in this case we instantiate it as empty, on purpose
    const range = document.createRange()
    // creates a mini-document (light weight version), in our range with our script in it
    const documentFragment = range.createContextualFragment(script)
    // appends it to our script root - so it renders in the correct location
    scriptRoot.current.append(documentFragment)
  }, [])

  return <div ref={scriptRoot} />
```

**Reply by snazzybytes** [reply]
interesting. which NextJS version did you have to do this in by the way? i stumbled into the same problem earlier and researching what's causing this in nextjs. did you do this in _app.tsx or layout.tsx (in case you wrap your pages in a layout like me). anyways ill figure it out but thanks

my situation was where Im loading a script on the bottom of a component (say even the wrapper layout.tsx), which controls a dialog box (open or close). 
i can click it 100 times and script works just fine... however as many others pointed when you use LINK and navigate to different internal page, the script stopped working. i guess it stems from my lack of understanding of nextjs framework internal ways and optimizations. cheers

**Reply by fvaldez421** [reply]
@snazzybytes - I believe this was on Next 12 (not using the app directory). That said, I was able to implement this on the component itself. In my case, it was a literal reusable script component.

**Reply by snazzybytes** [reply]
@fvaldez421 thanks friend. makes sense, i ended up doing what you suggested, just a custom component instead of loading a script which i have not fully stepped through actually (plus it is better that way for learning purposes 🫵 🫡 🫂)

**Reply by webprismdevin** [reply]
this was effective for loading custom code from Sanity CMS -> Remix as well. Going to have to go back to some old Next projects and implement this.

**Reply by billyberkouwer** [reply]
Brilliant solution @fvaldez421, thank you! Worked for my cross-origin iframe

**Reply by martiniucanastasia** [reply]
But what about SEO? I see you switched it to client side, but in my case I need the SEO and want to keep `<JsonLdScript/>` loaded on server.

**Comment by ari-motors-team** [comment]
I found a solution to UET. It seams to work, but please test yourself. This principle may apply to other scripts as well.
Regarding to UET and SPA in general, I found this.
https://bingadsuet.azurewebsites.net/SPA_Index.html

The solution in nextjs was adding this to `_app.js` `<Head />`.

```html
<script id="uetq">{`(function(w,d,t,r,u){....);`}</script>
```

and fire a page load event, when the page changes.

```js
export default function MyApp({ Component, pageProps }) {
  useEffect(() => {
    // uet
    if (typeof window !== "undefined" && window.uetq.push) {
      window.uetq = window.uetq || [];
      window.uetq.push("event", "page_view", {
        page_path: pageProps.page.path,
      });
    }
  }, [pageProps.page])
 ...
}
```

**Comment by vee2xx** [comment]
You can use useEffect to listen for the routeChangeComplete event and and reload (or load the portion you want loaded again) the script. Solved my issue 100%

Edit the _app.js file adding the following:

 ```
import Router from "next/router"
import {useEffect} from 'react';

 useEffect(() => {
    "use strict";
    Router.events.on("routeChangeComplete", (path) => {
      console.log(`Route change complete.`);
      setTimeout(() => {
      //Write the JS code here 
      }, 0);
      });

      return () => {
        Router.events.off('routeChangeComplete', 0);
    }
  }, [Router]);
```
Source of inspiration for this fix: https://stackoverflow.com/questions/68389552/how-to-access-the-url-or-page-that-is-currently-loading-on-navigation-with-next/68415175#68415175

NextJs developers should modify the <Script> loading library to add a strategy that will allow a new option something like onRouteChange that'll allow you to easily load a script when router path change event occur - this is a huge oversight and forcing lots of devs to give up using <Link> lib (which sucks because it actually is a nice thing that speeds up page loads).

**Comment by HallofFamer** [comment]
I found a solution that I can wrap your component in a HOC called withScript(Component, dir, ...srcs), which makes use of `useEffect` to append or remove scripts when a component mounts/unmounts: 

```
import {useEffect} from "react";

function withScript(Component, dir, ...srcs){
    function componentWithScriptProp(props){
        // eslint-disable-next-line react-hooks/rules-of-hooks
        srcs.forEach(src => {
            useEffect(() => {
                const script = document.createElement("script");
                script.src = `/${dir}/${src}.js`;
                script.defer = true;
                document.body.appendChild(script);
                return () => { document.body.removeChild(script) };
            });
        });
        return <Component {...props} />
    }
    componentWithScriptProp.getInitialProps = Component.getInitialProps;
    componentWithScriptProp.origGetInitialProps = Component.origGetInitialProps;
    if (process.env.NODE_ENV !== 'production') {
        const name = Component.displayName || Component.name || 'Unknown';
        componentWithScriptProp.displayName = `withScript(${name})`;
    }
    return componentWithScriptProp;
}

export default withScript
```

Now you just have to import `withScript` and wrap your component like this(assuming the component is called `MyComponent` and all scripts are located in folder `/public/js/`): 

```
// Use one script
export default withScript(MyComponent, "js", "tooltip")

// Use multiple script
export default withScript(MyComponent, "js", "tooltip", "popover", "datepicker")
```

Maybe its not an elegant solution, but it seems to serve the need for now. Maybe I can improve it so that it wont call `useEffect()` n times for n scripts, though I feel that its highly unlikely that you load more than 5 scripts at a time so I dont think it will cause noticeable slowdown.

**Comment by zackbiernat** [comment]
This is the workaround to get `<script>`s working on a nextjs app.
```
useEffect(() => {
        const script = document.createElement('script');

        script.src = 'https://example.com/somescript.js';
        script.async = true;

        document.body.appendChild(script);

        return () => {
            document.body.removeChild(script);
        };
    }, [propThatChanged]); // router prop or w/e
```

**Reply by saqib053** [reply]
This solution worked for me. you can replace propThatChange with 'router'

**Reply by zackbiernat** [reply]
I had a 3rd party script that would load and modify an element on screen and this solution did not work. If anyone experiences similar issues my hacky solution was to load the script (and element-to-be-modified) on a separate page and then point an iframe at it like so:
```                     
<iframe style={{ border: 'none' }} src={absolutePathToScriptPage} />
```

While script tags seem not to work within the context of next/router, the iframe is able to replicate a reload of the script.

**Reply by hoangvu12** [reply]
> This is the workaround to get `<script>`s working on a nextjs app.
> 
> ```
> useEffect(() => {
>         const script = document.createElement('script');
> 
>         script.src = 'https://example.com/somescript.js';
>         script.async = true;
> 
>         document.body.appendChild(script);
> 
>         return () => {
>             document.body.removeChild(script);
>         };
>     }, [propThatChanged]); // router prop or w/e
> ```

Thank you, saved me from countless hours.

**Reply by k6alvarez** [reply]
This is the solution that worked for me as well, thank you!!!

**Comment by beniusij** [comment]
Some proposed solutions work when refreshing a page in a browser, but not when navigating between the pages. That is still an issue and has been a problem for nearly 2 years.

**Reply by dmoKruso** [reply]
Same here... I couldn't solve the problem either, ended up just ditching the Link tags, and fall back to using normal a tag for navigation, that way we force to refresh the page on each navigation. 
But I am still hoping for a solution for this.

**Comment by nilwurtz** [comment]
I solved it this way.
`next/script` not working when routes changed.

```
  const router = useRouter()
  const [Js, setJs] = useState<React.ReactNode | null>(null)

  const handleComplete = () => {
    setJs(null)
    setJs(
      <Head>
        <script src={<url>} async />
      </Head>
    )
  }

  useEffect(() => {
    setJs(
      <Head>
        <script src={<url>} async />
      </Head>
    )
    router.events.on('routeChangeComplete', handleComplete)
  }, [router.events])

  return (
    <>
      {Js}
```

**Reply by AashirBhusal** [reply]
This does not work with Next useRouter ("next/navigation"). The useRouter("next/router") cannot be used anymore as it says router not mounted

**Comment by Akifcan** [comment]
You can give unique url for every render.
For example

```
<Script key={i} className='script-tag-js' src={`/path/to/script?v=${Math.random() * 999}`} type='module' />`
```

You can use this hook also.




```
import React, { ReactNode, useState, useEffect } from "react"
import { useRouter } from "next/router"
import Script from "next/script"

const useScript = (paths: string[]) => {
 const [scriptTag, setScriptTag] = useState<ReactNode>(<></>)
    const router = useRouter()


    useEffect(() => {
        if (!router.isReady) return

        router.events.on("routeChangeStart", () => {
            setScriptTag(<></>)
        })

        setScriptTag(<>
            {paths.map((x, i) => {
                return <Script key={i} className='script-tag-js' src={`${x}?v=${Math.random() * 999}`} type='module' />
            })}
        </>)


        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [router.asPath, router.isReady])

    return { scriptTag }

}

export default useScript

```

```
const Container: FC<ContainerProps> = ({ children, headerType, scripts = [] }) => {

    const { scriptTag } = useScript(['/assets/scripts/helpers/header.js', '/assets/scripts/helpers/modal.js', ...scripts])

    return <>
        <Header type={headerType} />
        {children}
        {scriptTag}
        <Footer />
    </>
}

export default Container
```

**Comment by IgorBadanyuk** [comment]
I've tried all the options and so far no results. Even mount and unmount Head with scripts. I don't want to remove Link component from the site because we are losing our speed advantage. 

Maybe someone will help. My case is displaying embedded posts from facebook and twitter. 

The principle is the same, connect next/script and subscribe to router event. The thing is, most of the scripts have a reload function in the documentation, so we call them when the route changes. 

```
const PageExternalScripts = () => {

const handleComplete = () => {
    if (window === undefined) {
      return
    }

    // Call FB SDK parse function
    if (window.FB !== undefined) {
      window.FB.XFBML.parse()
    }
    
    // Call Twitter widjet js function
    if (window.twttr !== undefined) {
      window.twttr.widgets.load()
    }
  }

  useEffect(() => {
    Router.events.on("routeChangeComplete", handleComplete)
  }, [])

  return (
    <>
      <Script
        src="https://platform.twitter.com/widgets.js"
        strategy="lazyOnload"
      />
      <div id="fb-root" />
      <Script
        src="https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v15.0"
        strategy="lazyOnload"
      />
    </>
  )

```

**Comment by RowinVanAmsterdam** [comment]
Unfortunately it seems like it's still an issue. I built a custom to hook for it and wrote a [blog post](https://rowinvanamsterdam.com/blog/load-scripts-per-page-while-using-next-link-for-navigation-in-nextjs) about it for further explanation. Hopefully it helps someone. 

The hook looks like this:
```
import { useEffect, useState } from 'react';

export const useExternalScript = (url: string) => {
    const [isLoaded, setIsLoaded] = useState(false);
    const [error, setError] = useState(false);

    useEffect(() => {
        const script = document.createElement('script');
        script.src = url;
        script.async = true;
        document.head.appendChild(script);

        script.onload = () => setIsLoaded(true);
        script.onerror = () => setError(true);

        return () => {
            document.head.removeChild(script);
        };
    }, [url]);

    return {
        isLoaded,
        error
    };
};
```

**Reply by harshalone** [reply]
this hook is not working error: 

```
Unhandled Runtime Error
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.

Source
app/hooks/useExternalScript.js (4:45) @ useExternalScript

  2 | 
  3 | export const useExternalScript = (url) => {
> 4 | const [isLoaded, setIsLoaded] = useState(false);
```

**Reply by RowinVanAmsterdam** [reply]
That’s because you’re calling the hook from a place that isn’t allowed. Please read the rules: https://reactjs.org/link/invalid-hook-call

**Reply by AashirBhusal** [reply]
hi what about inline internal js script, i am trying to integrate disqus to my website through universal disqus script. How should i use the hook to make it work?. the problem is i need next Link tag and cannot use anchor tag. The disqus script loads initially but if i route to next page then it stops loading and need to do hard refresh to load again. As i know this is done by link tag on purpose but is there any way to get around it.

**Comment by dshuthwal** [comment]
External <script> runs only on page load, not on page update. async and defer just for performance. To run the script on soft SPA you need to have some global bootstrap functions like window.init() in the script. You need to run bootstrap functions on script load. Full script cannot be run in soft SPA.

**Comment by oliver-daniel** [comment]
Necro'ing this thread to confirm it's still an issue in `v13.2.x` and beyond -- none of the fancy `useEffect` hacks worked to get a script with the same `src`, but different data parameters, to reload when navigating between pages with the same routing. The only thing that worked was changing all of the links available on one form page, which led to the others, back to `<a>` tags from `<Link>`s and forcing a hard reload on navigation.

**Reply by Bikbokken** [reply]
Thank you for sharing this! This works as intended!

**Comment by MuhammedQuresh** [comment]
Has anyone found a solution?

**Comment by MuhammedQuresh** [comment]
Found a solution


Create a scripts loaded component


`import React, { useEffect, useRef } from 'react';

const ScriptLoaded = () => {
  const scriptsRef = useRef([]);

  useEffect(() => {
    // Generate a new random value for cache busting
    const randomValue = Math.round(Math.random() * 100);

    // Define the list of script URLs to load
    const scriptUrls = [
      `/static/vendors/modernizr.min.js`,
      //`/static/vendors/jquery.min.js`,
      `/static/vendors/jquery-ui/jquery-ui.min.js`,
      `/static/vendors/bootstrap/js/bootstrap.min.js`,
      `/static/js/theme-vendors.js`,
      `/static/js/theme.min.js`,
      `/static/vendors/imagesloaded.pkgd.min.js`,
      `/static/vendors/anime.min.js`,
    ];

    // Create a promise for each script loading operation
    const loadScript = (src) => {
      return new Promise((resolve, reject) => {
        const script = document.createElement('script');
        script.src = src;
        script.async = true;
        script.onload = resolve;
        script.onerror = reject;
        document.body.appendChild(script);
        scriptsRef.current.push(script);
      });
    };

    // Load jQuery first and then load the other scripts
    loadScript(`/static/vendors/jquery.min.js`)
      .then(() => {
        return Promise.all(
          scriptUrls
            .filter((url) => url !== `/static/vendors/jquery.min.js`)
            .map((url) => loadScript(url))
        );
      })
      .catch((error) => {
        console.error('Error loading scripts:', error);
      });

    // Cleanup - Remove scripts loaded for the previous page
    return () => {
      scriptsRef.current.forEach((script) => {
        if (document.body.contains(script)) {
          document.body.removeChild(script);
        }
      });
      scriptsRef.current = [];
    };
  }, []);

  return <div />;
};

export default ScriptLoaded;
`



import the component in _app.js


`// pages/_app.js
import React, { useEffect } from 'react';
import { AppProps } from 'next/app';
import ScriptLoaded from '../components/script';

import 'public/static/vendors/liquid-icon/liquid-icon.min.css';
import 'public/static/vendors/font-awesome/css/font-awesome.min.css';
import 'public/static/css/theme-vendors.min.css';
import 'public/static/css/theme.min.css';
import 'public/static/css/themes/creative.css';
import 'public/static/css/hcl-fleet.css';
import router, { useRouter } from 'next/router';

function MyApp({ Component, pageProps }: AppProps) {
  const router = useRouter(); 
  return (
    <>
      <Component {...pageProps} />
      <ScriptLoaded key={router.asPath} />
    </>
  );
}

export default MyApp;
`

**Reply by NandaKyaw01** [reply]
Yeah it work!!!! And this is Nextjs 13 with app directory / app router
It reload js script files when route change with pathname useEffect parameter
```
'use client'
import { useEffect } from 'react'
import { usePathname } from 'next/navigation'

export default function ScriptLoaded() {
  const pathname = usePathname()
  useEffect(() => {
    // Generate a new random value for cache busting
    // const randomValue = Math.round(Math.random() * 100)

    // Define the list of script URLs to load
    const scriptUrls = [
      '/assets/js/vendor/modernizr-3.6.0.min.js',
      '/assets/js/vendor/jquery-3.6.0.min.js',
      '/assets/js/vendor/bootstrap.bundle.min.js',
      '/assets/js/plugins/slick.min.js',
      '/assets/js/plugins/countdown.min.js',
      '/assets/js/plugins/nice-select.min.js',
      '/assets/js/plugins/jqueryui.min.js',
      '/assets/js/plugins/image-zoom.min.js',
      '/assets/js/plugins/imagesloaded.pkgd.min.js',
      '/assets/js/plugins/ajaxchimp.js',
      '/assets/js/main.js',
    ]

    // Create a promise for each script loading operation
    const loadScript = (src: string) => {
      return new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = src
        script.async = true
        script.onload = resolve
        script.onerror = reject
        document.body.appendChild(script)
      })
    }

    // Load jQuery first and then load the other scripts
    loadScript('/assets/js/vendor/jquery-3.6.0.min.js')
      .then(() => {
        return Promise.all(
          scriptUrls
            .filter((url) => url !== '/assets/js/vendor/jquery-3.6.0.min.js')
            .map((url) => loadScript(url))
        )
      })
      .catch((error) => {
        console.error('Error loading scripts:', error)
      })
  }, [pathname])

  return null
}
```

and this is my layout file

```
import Footer from '@/components/footer'
import MainNavbar from '@/components/main-navbar'
import ScriptLoaded from '@/components/navigation-event'
import { navbarConfig } from '@/config/nav-bar'
import type { Metadata } from 'next'
import '../assets/css/plugins/animate.css'
import '../assets/css/plugins/jqueryui.min.css'
import '../assets/css/plugins/nice-select.css'
import '../assets/css/plugins/slick.min.css'
import '../assets/css/style.css'
import '../assets/css/vendor/bootstrap.min.css'
import '../assets/css/vendor/font-awesome.min.css'
import '../assets/css/vendor/pe-icon-7-stroke.css'

export const metadata: Metadata = {
  title: 'Jewelry Shop',
  description: 'You can buy jewelry as much as you want',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {

  return (
    <html lang="en">
      <head>
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1, shrink-to-fit=no"
        />
        <link
          href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,900"
          rel="stylesheet"
        />
      </head>
      <body>
        <MainNavbar items={navbarConfig.mainNav} />
        {children}
        <Footer />
        <ScriptLoaded />
      </body>
    </html>
  )
}
```

Thanks Brother

**Reply by rikkit** [reply]
This is the solution for me! My NextJS site was working fine apart from user journeys originating from 404 pages... where no external scripts were loaded at all. This approach now means scripts are loading on all my pages. Not sure why the Script tag works the way it does, it doesn't seem correct.

**Comment by douglasrcjames** [comment]
I have been running into this issue when loading iframes from external scripts such as a ticket buyer at DICE events and a SweepStakes iframe script from SweepWidget. This makes the <Script /> tag almost unusable on production because it gives such a terrible UX... none of the fixes above seem straight forward and I am worried it would just cause more issues. Currently my solution is to just present a button to reload the page if the user doesn't see anything. Would love to get the Script feature working properly with Next.js 13!

**Reply by NandaKyaw01** [reply]
I think Next Script is the problem with next router
Solution is to use vanilla script tag and 
Add or remove programmatically these script tag
Like I did , mine work when even route change
https://github.com/vercel/next.js/discussions/17919#discussioncomment-7020695

**Reply by douglasrcjames** [reply]
If you mean vanilla script as `<script />` element from standard HTML, then that's great and all but your code snippet you shared above is a ton of lines of code that should be achievable with the much more simple Next.js's `<Script />` component, which sounds like it doesn't work out of the box yet.

**Reply by NandaKyaw01** [reply]
Yeah I already test with Script tag and also with old pages route system and any other ways that I think of
But none of them work properly 
Yes js file does load, but when you navigate around you lose your third party js.
This is the only solution if you don't use a tag instead of nextjs Link tag
a tag reload the page  so what is the point of using react or nextjs right?


And also this is not a ton of code 
Just an old way of creating html tag using vanilla js
```
const loadScript = (src: string) => {
      return new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = src
        script.async = true
        script.onload = resolve
        script.onerror = reject
        document.body.appendChild(script)
      })
    }
```

**Reply by douglasrcjames** [reply]
gotcha, it seems to be working better, but am getting a `ReferenceError: document is not defined` in my server logs. Any idea how to fix that?

**Reply by douglasrcjames** [reply]
I think I misread the intention of this thread, and sounds like I should start another thread as this original issue was with the `<script />` tag not the next.js 13 `<Script />` tag!

However, @NandaKyaw01 your solution did help temporarily solve this, but only for one of my scripts that had a function used in the `onLoad` function (dice.fm tickets widget), and didn't work for my script that just ran an external script to render something in a div (SweepWidget)

**Reply by douglasrcjames** [reply]
Here's my new question if anyone else runs into a similar issue: https://github.com/vercel/next.js/discussions/55506

**Reply by NandaKyaw01** [reply]
Script not defined error
Reference error
I suppose you have many js files
One of your js file run before initialize

For example - your need to load j query file first in order to run my own script files

**Comment by akash-anadey** [comment]
I faced the same issue while I was implementing the AfterShip tracking widget. Here's the solution that worked for me:

```
useEffect(() => {
        if (!router.isReady) return

        const script = document.createElement('script');

        script.src = 'https://button.aftership.com/all.js?username=XXXXXXXX';
        script.async = true;

        document.body.appendChild(script);

        return () => {
            document.body.removeChild(script);
        };
}, [router.asPath, router.isReady]); 
```

**Comment by Merton** [comment]
When using Next app router and server rendered components, I found this approached fixed the issue:
1. Wrap your script / embed logic into a component, and import that into your page.
2. Keep the page as a server component (ie, no `use client`)
3. For the script component, this needs to be a client rendered component, so add `use client` at the top. And add @Akifcan 's random `Math.random()` to your src.

Mini example which fetches data server side:

```typescript
// Page.tsx
async function getData(): Promise<any> {
  const data = await fetch(...)

  return data
}

export default async function MyPage() {
  const data = await getData();

  return (
    <div>
        <div>
            <MyScriptComponent />
        </div>
    </div>
  );
}

// MyScriptComponent.tsx
'use client';

import Script from 'next/script';

export const MyScriptEmbed = () => {
  return (
    <div>
      <div id="myEmbedDiv"></div>
     <Script src={'path/or/url/to/script.js?v=' + Math.random() * 999} />
    </div>
  );
};

```

**Reply by Freundschaft** [reply]
while this works in general, I have the issue, that the script doesnt re-load when navigating, due to the router cache, any idea how to fix this?

**Reply by fvaldez421** [reply]
@Freundschaft - I found a solution for this above:
https://github.com/vercel/next.js/discussions/17919#discussioncomment-6995031

**Reply by Freundschaft** [reply]
thanks! how do you prevent the script from running and loading the iframe multiple times when the useeffect hook is invoked multiple times?

**Reply by Freundschaft** [reply]
ok I think I found a solution, im writing to a global variable in the window object and unsetting it in the cleanup logic of the useeffect hook, but with a settimeout of 100 ms, that does the trick, but its a bit hacky :/
